drv/lpc55-update-server/src/main.rs

```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
+                // a flash slot digest even though the contained image is
+                // the same.
+            }
+            Err(e) => {
+                return Err(e.into()); // Non-fatal error. We did not alter stage0.
+            }
+        }
+
+        // Finish by erasing the unused portion of flash bank.
+        // An error here means that the stage0 slot may not be clean but at least
+        // it has the intended bootloader written.
+        // TODO: There should be a unique error here as well as anywhere else
+        // in the bootloader update flow.
+        let erase_start = stage0.0.start.checked_add(len as u32).unwrap_lite();
+        self.flash_erase_range(erase_start..stage0.0.end)?;
+        Ok(())
+    }
+
+    fn flash_erase_range(
+        &mut self,
+        span: Range<u32>,
+    ) -> Result<(), UpdateError> {
+        // It's assumed that the caller has done safe math and that
+        // there is no danger here.
+        let word_span = (span.start / (BYTES_PER_FLASH_WORD as u32))
+            ..=(span.end / (BYTES_PER_FLASH_WORD as u32) - 1);
+
+        self.flash.start_erase_range(word_span);
+        loop {
+            match self.flash.poll_erase_or_program_result() {
+                None => continue,
+                Some(Ok(())) => return Ok(()),
+                Some(Err(_)) => return Err(UpdateError::FlashError),
+            }
+        }
+    }
+
+    fn compare_cache_to_flash(
+        &self,
+        span: &Range<u32>,
+    ) -> Result<(), UpdateError> {
+        // Is there a cached image?
+        // no, return error
+
+        // Lengths are rounded up to a flash page boundary.
+        let clen = self.valid_cache_image_present()?;
+        let flen = self.valid_flash_image_present(span)?;
+        if clen != flen {
+            return Err(UpdateError::ImageMismatch);
+        }
+        // compare flash page to cache
+        let cached =
+            self.fw_cache[0..flen / core::mem::size_of::<u32>()].as_bytes();
+        let mut flash_page = [0u8; BYTES_PER_FLASH_PAGE];
+        for addr in (0..flen).step_by(BYTES_PER_FLASH_PAGE) {
+            let size = if addr + BYTES_PER_FLASH_PAGE > flen {
+                flen - addr
+            } else {
+                BYTES_PER_FLASH_PAGE
+            };
+
+            indirect_flash_read(
+                &self.flash,
+                addr as u32,
+                &mut flash_page[..size],
+            )?;
+            // XXX fw_cache is [u32; FW_CACHE_MAX]
+            if flash_page[0..size] != cached[addr..addr + size] {
+                return Err(UpdateError::ImageMismatch);
+            }
+        }
+        Ok(())
+    }
+
+    // Looking at a region of flash, determine if there is a possible NXP
+    // image programmed. Return the length in bytes of the flash pages
+    // comprising the image including padding to fill to a page boundary.
+    fn valid_flash_image_present(
```

Comment:
nit, take it or leave it: it feels a bit strange to me that a method name ending in `_present` returns a number and not a boolean? `_present` feels like a bool to me. but, i'm open to being convinced otherwise.